sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -515,3 +515,96 @@ case class StringToMap(text: Expression, pairDelim: Expression, keyValueDelim: E
 
   override def prettyName: String = "str_to_map"
 }
+
+/**
+ * Adds/replaces field in struct by name.
+ *
+ * @param children Seq(struct, name, val)
+ */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(struct, name, val) - Adds/replaces field in struct by name.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(NAMED_STRUCT("a", 1, "b", 2), "c", 3);
+       {"a":1,"b":2,"c":3}
+      > SELECT _FUNC_(NAMED_STRUCT("a", 1, "b", 2), "b", 3);
+       {"a":1,"b":3}
+      > SELECT _FUNC_(CAST(NULL AS struct<a:int,b:int>), "c", 3);
+       {"a":null,"b":null,"c":3}
+      > SELECT _FUNC_(NAMED_STRUCT('a', 1, 'b', 2, 'b', 3), 'b', 100);
+       {"a":1,"b":100,"b":100}
+      > SELECT _FUNC_(a, 'a.c', 3) FROM (VALUES (NAMED_STRUCT('a', NAMED_STRUCT('a', 1, 'b', 2))) AS T(a));
+       {"a":{"a":1,"b":2,"c":3}}
+  """)
+// scalastyle:on line.size.limit
+case class WithField(children: Seq[Expression]) extends Unevaluable {
+  private lazy val Seq(structExpr, nestedAttributeExpr, valExpr) = children
+  private lazy val nestedAttribute = nestedAttributeExpr.eval().toString
+  private lazy val attributes = UnresolvedAttribute.parseAttributeName(nestedAttribute)
+  lazy val toCreateNamedStruct: Expression = toCreateNamedStruct(structExpr, attributes)
+
+  override def dataType: StructType = toCreateNamedStruct.dataType.asInstanceOf[StructType]
+
+  override def foldable: Boolean = toCreateNamedStruct.foldable
+
+  override def nullable: Boolean = toCreateNamedStruct.nullable
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val structTypeName = StructType(Nil).typeName
+    if (children.size != 3) {
+      TypeCheckResult.TypeCheckFailure(s"$prettyName expects 3 arguments.")
+    } else if (!structExpr.dataType.isInstanceOf[StructType]) {
+      TypeCheckResult.TypeCheckFailure(s"Only $structTypeName is allowed to appear at " +
+        s"first position, got: ${structExpr.dataType.typeName}.")
+    } else if (!(nestedAttributeExpr.foldable && nestedAttributeExpr.dataType == StringType)) {
+      TypeCheckResult.TypeCheckFailure(s"Only foldable ${StringType.catalogString} expressions " +
+        s"are allowed to appear at second position.")
+    } else if (nestedAttribute == null) {
+      TypeCheckResult.TypeCheckFailure("Field name should not be null.")
+    } else {
+      val structType = structExpr.dataType.asInstanceOf[StructType]
+      val intermediateAttributes :+ _ = attributes
+      intermediateAttributes
```

Comment:
can we reuse `StructType.findNestedField`?

Reply:
I use `StructType.findNestedField` 2 lines below this one. Not sure how I would reuse `StructType.findNestedField` for this line though. Could you be more specific?